### ISSUE TITLE
PL TTS:  numbers wrongly declined

### ISSUE THREAD

**Issue by bumbaras** [osmandapp/OsmAnd#7570]
How text to speech mechanism work? Is it some text file for given language with list of all words and the engine TtS is doing rest? In Polish language the lecteur tries very hard to declined street numbers by whatever he/she can (which is totally wrong) which sounds funny at first but after some time it stops to be funny and becomes annoying. Is it something normal user can fix by himself? For example by editing some text files?

**Comment by bumbaras** [comment]
How text to speech mechanism work? Is it some text file for given language with list of all words and the engine TtS is doing rest? In Polish language the lecteur tries very hard to declined street numbers by whatever he/she can (which is totally wrong) which sounds funny at first but after some time it stops to be funny and becomes annoying. Is it something normal user can fix by himself? For example by editing some text files?

**Comment by mikehgentry** [comment]
#7379 sounds related - you might get better results by changing TTS engine in the OS.

**Comment by sonora** [comment]
It will be tough for me to help, but I can try. First we need to establish if it is an issue of the TTS engine you are using, or in deed the OsmAnd config file which contains the grammar.

For a first test, can you try several different TTS engines and check if they are all consistently wrong?

For this, and as a second test, you can enable the OsmAnd Development plugin, then go to its settings, and try the test buttons. Tell me which ones sound wrong (expkain what they say), and state what you expect.

**Comment by varabiusz** [comment]
I can confirm this issue. And yes, it is annoying. :P Not only road numbers, but also driving time.
I checked the test messages:

**1.1**

There is:
... Przewidywany czas to cztery **minut**.

Expected:
... Przewidywany czas to cztery **minuty**.

1.2 - ok

**1.3**

There is:
... Przewidywany czas to **dwie godzin, dwóch minut.**

Expected:
... Przewidywany czas to **dwie godziny, dwie minuty.**

1.4
There is:
... Przewidywany czas to **trzy godzin, trzech minut.**

Expected:
... Przewidywany czas to **trzy godziny, trzy minuty.**

1.5 - ok

2.1

There is:
Nowa trasa wyznaczona **dwunastu** kilometrów, przewidywany czas to pięć godzin **dziesięciu** minut.

Expected:
Wyznaczona nowa trasa to **dwanaście** kilometrów, przewidywany czas to pięć godzin **dziesięć** minut.

2.2

There is:
Nowa trasa wyznaczona, **dwudziestu** kilometrów, przewidywany czas to siedem godzin **dwudziestu pięciu** minut.

Expected:
Wyznaczona nowa trasa to **dwadzieścia** kilometrów, przewidywany czas to siedem godzin **dwadzieścia pięć** minut.

2.3

There is:
Nowa trasa wyznaczona, **dziewięćdziesięciu** kilometrów, przewidywany czas to piętnaście godzin **trzydziestu** minut.

Expected:
Wyznaczona nowa trasa to, **dziewięćdziesiąt** kilometrów, przewidywany czas to piętnaście godzin **trzydzieści** minut.

2.4

There is:
Nowa trasa wyznaczona, **stu dwudziestu sześciu** kilometrów, przewidywany czas to dwadzieścia pięć godzin **czterdziestu pięciu** minut.

Expected:
Wyznaczona nowa trasa to, **sto dwadzieścia sześć** kilometrów, przewidywany czas to piętnaście godzin **czterdzieści pięć** minut.

3.1

There is:
Za około **dwóch kilometrów** skręć lekko w lewo.

Expected:
Za około **dwa kilometry** skręć lekko w lewo.

3.2 - ok
3.3 - ok

3.4
There is:
Za około **trzech kilometrów** skręć w prawo w SR **osiemdziesięciu** w kierunku Rome.

Expected
Za około **trzy kilometry** skręć w prawo w SR **osiemdziesiąt** w kierunku Rome.

3.5
There is:
... Route **dwudziestu trzech**...
Expected:
... Route **dwadzieścia trzy**...

3.6 - ok
3.7

There is:
... zjazd sześciu ...
Expected:
... zjazd sześć ...

And so on...

**Comment by sonora** [comment]
From what I see, your observations result from two reasons:
1. OsmAnd's pl_tts.js file, which contains the word order and grammar rules for the voice prompts, does not seem to properly account for Polish grammar. Here is the vocabulary section regarding time: 
> 	// TIME SUPPORT
	dictionary["time"] = tts ? ", przewidywany czas to" : "time.ogg";
	dictionary["1_hour"] = tts ? "jedna godzina" : "1_hour.ogg";
	dictionary["hours"] = tts ? "godzin" : "hours.ogg";
	dictionary["less_a_minute"] = tts ? "mniej niż minuta" : "less_a_minute.ogg";
	dictionary["1_minute"] = tts ? "jedna minuta" : "1_minute.ogg";
	dictionary["minutes"] = tts ? "minut" : "minutes.ogg";


and as you can see it does e.g. not contain minut**y** or godzin**a** at all, that's why these words are never spoken.

2. Secondly, it appears the TTS engine you are using does not correctly form the numbers in the fashion they are grammatically needed in each sentence. From what I see OsmAnd only passes the digits, e.g. "2", to the TTS engine, but there is no code to determine when **dwóch** and when **dwa** would be required.

In summary I would say the OsmAnd pl_tts voice file needs a major upgrade to reflect correct Polish grammar, and it seems nobody has stepped forward to do this.

If you can summarize the grammar rules, I could maybe take a shot at coding this, but it would be a lot easier for someone in command of the Polish language. I think (1.) can be fixed easier, (2.) may be very challenging, because we cannot hard-code all numbers. Once we know the grammar rules in question here, maybe there is a similar language where the coding is already implemented? In that case I could try a simple transfer to the pl_tts file.

**Comment by varabiusz** [comment]
Polish language is super cool, so it won't be easy. ;)

Long story short:

1) Variation of a noun

number of | kilometer | hour | minute
---- | ---- | ---- | ----
1 | kilometr | godzina | minuta
2-4 | kilometry | godziny | minuty
5-21 | kilometrów | godzin | minut

Any value ending in 2-4 except those where the penultimate digit is 1 will be "kilometry", "godziny", "minuty" eg.:

22, 23, 24,
32, 33, 34,
...
92, 93, 94,
102, 103, 104,
122, 123, 124,
...
1002, 1003, 1004,
1022, 1023, 1024,
1122, 1123, 1124
...

The examples below will be "kilometrów", "godzin", "minut" as in all other cases.

12, 13, 14,
112, 113, 114,
1012, 1013, 1014,
...


2) In polish language there are gender types for which the number is conjugated differently:

kilometer - masculine gender
hour, minute - feminine gender


number of | kilometer | hour / minute
---- | ---- | ----
1 | jeden | jedna
2 | dwa | dwie

Other values are identical to simple numerals (not ordinal numbers!) except for those ending in 2 if there is no 1 before, eg.:

number of | kilometer | hour / minute
---- | ---- | ----
22 | dwadzieścia dwa | dwadzieścia dwie
52 | pięćdziesiąt dwa | pięćdziesiąt dwie


A more detailed description of the rules can be found here:
https://polski.info/en/grammar/numerals

**Comment by sonora** [comment]
Great summary and link, thanks.

I am wondering if some or maybe all TTS engines capable of supporting Polish can determine the correct declination if we use nit abbreviations instead of fixed vocabulary. Can I ask you to test this file: [pl_tts.zip](https://github.com/osmandapp/OsmAnd/files/6997768/pl_tts.zip)

You need to unzip it and place it in your device's OsmAnd data folder in sub-folder `voice/pl-tts`. Before that please rename the existing `pl-tts.js` there. Please make sure you use the active OsmAnd folder, there may be several in internal and external or SD card memory. Then restart OsmAnd.

**Comment by varabiusz** [comment]
I made a test using the file after your change. It is better. Minutes and kilometers are OK - the problem with hours remained. I checked all messages:

1.1 - ok
1.2 - ok
1.3
There is:
... Przewidywany czas to **dwa h**, dwie minuty.
Expected:
... Przewidywany czas to **dwie godziny**, dwie minuty.
1.4
There is:
... Przewidywany czas to **trzy h**, trzy minuty.
Expected:
... Przewidywany czas to **trzy godziny**, trzy minuty.
1.5 - ok

2.1
There is:
Nowa trasa wyznaczona dwunastu kilometrów, przewidywany czas to **pięć h** dziesięć minut.
Expected:
Wyznaczona nowa trasa to dwanaście kilometrów, przewidywany czas to **pięć godzin** dziesięć minut.
2.2
There is:
Nowa trasa wyznaczona, dwadzieścia kilometrów, przewidywany czas to **siedem h** dwadzieścia pięciu minut.
Expected:
Wyznaczona nowa trasa to dwadzieścia kilometrów, przewidywany czas to **siedem godzin** dwadzieścia pięć minut.
2.3
There is:
Nowa trasa wyznaczona, dziewięćdziesiąt kilometrów, przewidywany czas to **piętnaście h** trzydziestu minut.
Expected:
Wyznaczona nowa trasa to, dziewięćdziesiąt kilometrów, przewidywany czas to **piętnaście godzin** trzydzieści minut.
2.4
There is:
Nowa trasa wyznaczona, sto dwadzieścia sześć kilometrów, przewidywany czas to **dwadzieścia pięć h** czterdziestu pięciu minut.
Expected:
Wyznaczona nowa trasa to, sto dwadzieścia sześć kilometrów, przewidywany czas to **dwadzieścia pięć godzin** czterdzieści pięć minut.

3.1 - ok
3.2 - ok
3.3 - ok
3.4
There is:
Za około **trzech kilometrów** skręć w prawo w SR osiemdziesięciu w kierunku Rome.
Expected
Za około **trzy kilometry** skręć w prawo w SR osiemdziesiąt w kierunku Rome.
3.5 (in street names shoud be use cardinal numbers)
There is:
... Route **dwudziestu trzech**...
Expected:
... Route **dwadzieścia trzy**...
3.6 - ok
3.7
There is:
Skręć lekko w prawo **w zjazd sześciu** ...
Expected:
Skręć lekko w prawo **w szósty zjazd** ...
or
Skręć lekko w prawo w zjazd sześć ...
3.8
There is:
Za trzysta pięćdziesiąt metrów skręć lekko w prawo **w zjazd sześciu** ...
Expected:
Za trzysta pięćdziesiąt metrów skręć lekko w prawo w zjazd sześć ...
or
Za trzysta pięćdziesiąt metrów skręć lekko w prawo **w szósty zjazd** ...
3.9 - like above
3.10 - like above
3.11
There is:
Skręć lekko w prawo **w zjazd dwustu sześćdziesięciu B**
Expected:
Skręć lekko w prawo **w zjazd dwieście sześciedsiąt B**
3.12
There is:
Skręć w lewo **w zjazd piętnastu B**
Expected:
Skręć w lewo **w zjazd piętnaście B**

4.1 - ok
4.2 - ok
4.3 - ok
4.4 - ok

5.1 - ok
5.1 - almost OK - "I 15" -> there is: "i piętnastu", expected: "i piętnaście".
5.2 - almost OK - "Highway 60" -> there is "sześćdziesięciu", exprcted: "sześćdziesiąt"

6.1 - ok
6.2 - ok
6.3 - ok
6.4 - ok

7.1 - ok
7.2 - ok
7.3 - ok
7.4 - ok
7.5 - ok
7.6 - ok
7.7 - ok

8.1 - ok
8.2 - ok
8.3 - ok
8.4 - ok
8.5 - almost OK - "Mijasz POI" -> "Mijasz punkt"?

9.1 - ok
9.2 - ok
9.3 - ok
9.4 - ok
9.5 - ok
9.6 - ok
9.7 - ok
9.8 - ok
9.9 - ok

10.1 - ok
10.2 - ok
10.3 - ok
10.4 - ok

**Comment by varabiusz** [comment]
I changed "h" to "godz":

`dictionary["hours"] = tts ? "godz" : "hours.ogg";`

and after retest:

1.3 - ok 
1.4 - ok
2.1 - ok 
2.2 - ok
2.3 - ok
2.4 - ok

:)

**Comment by sonora** [comment]
Ok, thanks for the good news! I have now updated the repository with my idea, see https://github.com/osmandapp/OsmAnd-resources/commit/83977cb240d975b723bb86360acc7836d8b7ff45.

So what's still open now? 

- [ ] Looks like the **3.x** commands still have some remaining "kilometer" issues? But I have no immediate idea how to improve further.

- [ ] Looks like highway numbers in **5.1/5.2** are still not perfect, but also here we would need a new idea?

- [ ] Regarding using "_punkt_" for "_POI_", is that a general request? There are 2 occurrences of "POI", `mijasz POI` and `następnie miniesz POI`, would need correct wording for both. We can also space out as "P O I", in which case most engines would spell the letters separately.

**Comment by varabiusz** [comment]
There are still a few issues:

1. POI - 7,7 and 8.5

TTS voice reads it as "poi" - there is no such word in Polish. POI is an acronym in English, so it has no Polish meaning either. It should be translated, but (lest it be too easy), there is no direct translation. :)

I think the most relevant translation is "punkt użyteczności publicznej" or even simpler "użyteczne miejsce" (useful place):

```
dictionary["and_arrive_poi"] = tts ? "następnie miniesz użyteczne miejsce  -" : "and_arrive_poi.ogg";
dictionary["reached_poi"] = tts ? "mijasz użyteczne miejsce -" : "reached_poi.ogg";
```

2. A little thing...

Let's add a pause at the end of:
```
dictionary["route_calculate"] = tts ? "Nowa trasa wyznaczona" : "route_calculate.ogg";
```
after change:
```
dictionary["route_calculate"] = tts ? "Nowa trasa wyznaczona," : "route_calculate.ogg";
```


3. Incorrect form of "kilometers"

3.4
There is:
Za około **trzech kilometrów** skręć w prawo w SR osiemdziesięciu w kierunku Rome.
Expected
Za około **trzy kilometry** skręć w prawo w SR osiemdziesiąt w kierunku Rome.

This is the only case incorrect declination of "kilometer". I have no idea how this can be fixed.


4. Incorrect form of numerals in street names:

3.4 (SR 80)
There is:
Za około trzech kilometrów skręć w prawo w SR **osiemdziesięciu** w kierunku Rome.
Expected
Za około trzy kilometry skręć w prawo w SR **osiemdziesiąt** w kierunku Rome.

3.5 (Route 23)
There is:
Za trzysta pięćdziesiąt metrów skręć w lekko w prawo w Route **dwudziestu trzech** Main Street, następnie trzymaj się lewej strony.
Expected:
Za trzysta pięćdziesiąt metrów skręć w lekko w prawo w Route **dwadzieścia trzy** Main Street, następnie trzymaj się lewej strony.

5.2 (I 15)
There is:
Za czterysta pięćdziesiąt metrów wjedź na rondo i wybierz pierwszy zjazd w "**I piętnastu**" w kierunku Los Angeles
Expected:
Za czterysta pięćdziesiąt metrów wjedź na rondo i wybierz pierwszy zjazd w "**I piętnaście**" w kierunku Los Angeles

5.3 (Highway 60)
There is:
Wybierz drugi zjazd w Highway **sześćdziesięciu**.
Expected:
Wybierz drugi zjazd w Highway **sześćdziesiąt**.


5. Incorrect form of numerals in exits

3.7, 3.8, 3.9, 3.10 (onto exit 6)
There is:
... w zjazd **sześciu** ...
Expected:
... w **szósty** zjazd ...
or
... w zjazd **sześć** ...

3.11 (to exit 260B)
There is:
... w zjazd  **dwustu sześćdziesięciu B**
Expected:
... w zjazd  **dwieście sześćdziesiąt B**

3.12 (to exit 15B)
There is:
... w zjazd **piętnastu B**
Expected:
... w zjazd **piętnaście B**


See the **getExitNumber** method. Change the line:
````
if (!tts && exitInt > 0 && exitInt < 18)
````
to:
````
if (exitInt > 0 && exitInt < 18)
````
This will fix the issues in 3.7 - 3.10


6. I found a comment that says "left_keep"/"right_keep" refers to the aiding lane selection of the roadway. I thought something was wrong with messages 4.x...

If so... this:
````
dictionary["left_keep"] = tts ? "trzymaj się lewej strony" : "left_keep.ogg";
dictionary["right_keep"] = tts ? "trzymaj się prawej strony" : "right_keep.ogg";
````	
should be changed to:
````
dictionary["left_keep"] = tts ? "trzymaj się lewego pasa" : "left_keep.ogg";
dictionary["right_keep"] = tts ? "trzymaj się prawego pasa" : "right_keep.ogg";
````
Then these messages will start to make more sense. ;)

And your commit with the previous changes is OK, thanks.

**Comment by sonora** [comment]
Ok, I have implemented your feedback regarding

- [x] **1:** "POI", test buttons 7,7 and 8.5
- [x] **2:** "comma for tts after 'route_calculate' "
- [x] **5a:** "handling of numerical exits"
- [x] **6:** "left_keep/right_keep" wording adjustment

For

- [ ] **3:** the correct declination of the numbers in connection with "kilometers", please keep in mind that we now simply pass "km" to the TTS engine and rely on proper handling,which solved a number of related issues
- [ ] **4:** the correct declination of the numbers in connection with "highway numbers", we entirely rely on the TTS engine handling this
- [ ] **5b:** "handling of non-numerical exits", see 4. above

For all these we just pass the text/number combinations to the TTS engine, which handles the rest. I think https://github.com/osmandapp/OsmAnd/issues/7379#issuecomment-520219615 applies accordingly: Maybe update your TTS engine to the latest version, and/or try a few different built-in or third party TTS engines, they all differ in ability and prioritization of how to pronounce things. Maybe you find one which handles navigation instructions better than others which may be more focused on scientific or mathematical pronunciation.

Changes are committed to the repo, here is a copy for testing purposes: [pl_tts.zip](https://github.com/osmandapp/OsmAnd/files/7072498/pl_tts.zip)

**Comment by varabiusz** [comment]
Hi,
Thank you for changes, but... There are some regresions after change you made in commit [a436446](https://github.com/osmandapp/OsmAnd-resources/commit/a436446)

You shouldn't comment lines 319 and 320. Now there is no exit number in message 3.11:

It is:
```
Skręć lekko w prawo w zjazd
```
Expected:
```
Skręć lekko w prawo w zjazd 260B
```

Line 317 and 322 are the same, so what is the sense of condition - if there is TSS or if doesn't - the same code will be executed.
You should remove "!tts" from line 316 and replace it by condition that checks if exit number not contains string. This extra condition is necessary because in 3.12 there is no "B" at the ond of exit number.

What values are for exitString and exitInt when you test 3.11 message? Are you able to check it?

**Comment by sonora** [comment]
Ah, yes, I had misunderstood you. Should be fixed now, I also updated the pl_tts.zip linked in my comment above for easier testing.

The workaround is a bit awkward, because as you can test now
- button 3.11 with "Exit 260B" is processed via the _exitString_ as "260B", probably producing the wrong number declination on at least some TTS engines
- but button 3.12 with "Exit 15B" is processed as explicit integer _nth(exitInt)_, using the file's built in manual declination (the one also used for the recorded cases). This leaves out the "B"information, because we fork to the integer-only case.

Is that how you think you want it for now?

**Comment by varabiusz** [comment]
Your commit [9e66a06](https://github.com/osmandapp/OsmAnd-resources/commit/9e66a06) is almost correct ;)

You omitted the second part of my comment about extra condition which will test whether the exit number contains a string.
Then 3.12 will not leaves out the "B" information.

This should do the trick:
```
if (exitInt > 0 && exitInt < 18 && exitInt.toString() === exitString) { }
```

**Comment by sonora** [comment]
Yes, I like that solution better, too: In the TTS case rather have bad grammar but full info than leaving out info. In the recorded case we can leave the fallback to at least pronounce the numbers, I guess.

**Comment by pm4rcin** [comment]
Hi, the open source RHVoice has a problem with "implicit" kilometers and meters but also hours and minutes. Basically for it to work correctly you'd have to revert https://github.com/osmandapp/OsmAnd-resources/commit/83977cb240d975b723bb86360acc7836d8b7ff45. The interesting thing is that test 5.1 works correctly. Most of them however fails. I hope you understood and if not let me know and I'll explain in detail.

**Comment by akotynski** [comment]
> Hi, the open source RHVoice has a problem with "implicit" kilometers and meters but also hours and minutes. Basically for it to work correctly you'd have to revert [osmandapp/OsmAnd-resources@83977cb](https://github.com/osmandapp/OsmAnd-resources/commit/83977cb240d975b723bb86360acc7836d8b7ff45). The interesting thing is that test 5.1 works correctly. Most of them however fails. I hope you understood and if not let me know and I'll explain in detail.

Hey, it should be fixed with https://github.com/osmandapp/OsmAnd-resources/pull/907

**Comment by akotynski** [comment]
> There are still a few issues:
> 
> 1. POI - 7,7 and 8.5
> 
> TTS voice reads it as "poi" - there is no such word in Polish. POI is an acronym in English, so it has no Polish meaning either. It should be translated, but (lest it be too easy), there is no direct translation. :)
> 
> I think the most relevant translation is "punkt użyteczności publicznej" or even simpler "użyteczne miejsce" (useful place):
> 
> ```
> dictionary["and_arrive_poi"] = tts ? "następnie miniesz użyteczne miejsce  -" : "and_arrive_poi.ogg";
> dictionary["reached_poi"] = tts ? "mijasz użyteczne miejsce -" : "reached_poi.ogg";
> ```
> 
> 2. A little thing...
> 
> Let's add a pause at the end of:
> 
> ```
> dictionary["route_calculate"] = tts ? "Nowa trasa wyznaczona" : "route_calculate.ogg";
> ```
> 
> after change:
> 
> ```
> dictionary["route_calculate"] = tts ? "Nowa trasa wyznaczona," : "route_calculate.ogg";
> ```
> 
> 3. Incorrect form of "kilometers"
> 
> 3.4 There is: Za około **trzech kilometrów** skręć w prawo w SR osiemdziesięciu w kierunku Rome. Expected Za około **trzy kilometry** skręć w prawo w SR osiemdziesiąt w kierunku Rome.
> 
> This is the only case incorrect declination of "kilometer". I have no idea how this can be fixed.
> 
> 4. Incorrect form of numerals in street names:
> 
> 3.4 (SR 80) There is: Za około trzech kilometrów skręć w prawo w SR **osiemdziesięciu** w kierunku Rome. Expected Za około trzy kilometry skręć w prawo w SR **osiemdziesiąt** w kierunku Rome.
> 
> 3.5 (Route 23) There is: Za trzysta pięćdziesiąt metrów skręć w lekko w prawo w Route **dwudziestu trzech** Main Street, następnie trzymaj się lewej strony. Expected: Za trzysta pięćdziesiąt metrów skręć w lekko w prawo w Route **dwadzieścia trzy** Main Street, następnie trzymaj się lewej strony.
> 
> 5.2 (I 15) There is: Za czterysta pięćdziesiąt metrów wjedź na rondo i wybierz pierwszy zjazd w "**I piętnastu**" w kierunku Los Angeles Expected: Za czterysta pięćdziesiąt metrów wjedź na rondo i wybierz pierwszy zjazd w "**I piętnaście**" w kierunku Los Angeles
> 
> 5.3 (Highway 60) There is: Wybierz drugi zjazd w Highway **sześćdziesięciu**. Expected: Wybierz drugi zjazd w Highway **sześćdziesiąt**.
> 
> 5. Incorrect form of numerals in exits
> 
> 3.7, 3.8, 3.9, 3.10 (onto exit 6) There is: ... w zjazd **sześciu** ... Expected: ... w **szósty** zjazd ... or ... w zjazd **sześć** ...
> 
> 3.11 (to exit 260B) There is: ... w zjazd **dwustu sześćdziesięciu B** Expected: ... w zjazd **dwieście sześćdziesiąt B**
> 
> 3.12 (to exit 15B) There is: ... w zjazd **piętnastu B** Expected: ... w zjazd **piętnaście B**
> 
> See the **getExitNumber** method. Change the line:
> 
> ```
> if (!tts && exitInt > 0 && exitInt < 18)
> ```
> 
> to:
> 
> ```
> if (exitInt > 0 && exitInt < 18)
> ```
> 
> This will fix the issues in 3.7 - 3.10
> 
> 6. I found a comment that says "left_keep"/"right_keep" refers to the aiding lane selection of the roadway. I thought something was wrong with messages 4.x...
> 
> If so... this:
> 
> ```
> dictionary["left_keep"] = tts ? "trzymaj się lewej strony" : "left_keep.ogg";
> dictionary["right_keep"] = tts ? "trzymaj się prawej strony" : "right_keep.ogg";
> ```
> 
> should be changed to:
> 
> ```
> dictionary["left_keep"] = tts ? "zajmij lewy pas" : "left_keep.ogg";
> dictionary["right_keep"] = tts ? "zajmij prawy pas" : "right_keep.ogg";
> ```
> 
> Then these messages will start to make more sense. ;)
> 
> And your commit with the previous changes is OK, thanks.

Created PR to address this one https://github.com/osmandapp/OsmAnd-resources/pull/908

**Comment by sonora** [comment]
@akotynski Have you sufficiently tested it? Then I would be happy to merge it, of course!

**Comment by sonora** [comment]
PS: Can I please ask to test and make sure the new config file does not break any commands when used with recorded Polish voice. In particular, functions retuning things like`": "` break recorded commands if they are used in the recorded case (which for your changes may not be the case).

(To ease maintenance, I always use the same config file for TTS and recorded, in particular this also easier facilitates to port refinements we make in TTS also to recorded if anybody spares the effort.)

**Comment by akotynski** [comment]
Thanks!

Recordings for PL are pretty bad but I think I should not break anything... ;)  I changed `": "` to `" : "` because file `around.ogg:` was missing :D 
Actually, to support PL MediaPlayer with new logic it is enough to generate new recordings with proper grammar form but I would need some guidance on how to do it. Not sure if it's worth it, is anyone still using the recordings-based voice router? :)

Also, I went through all the test commands with Google TTS engine, I found one more mistake in 1.4 (and fixed it), everything else seems to work quite well.
I'm still a little bit not sure about commands 3.7, 3.8, 3.9, 3.10 (onto exit 6)
There is:
... w szósty zjazd ... (it's like take 6th exit, ex. on the roundabout)
I think it should be:
... w zjazd sześć ... (take exit no. 6)


I'm also thinking about one small improvement, in Poland we do not say: `one hour`, just `hour`. It's not a mistake but seems a little bit unnatural. Similarly for minutes, but if `one hour one minute` is left we would say `hour and one minute` so it's not black or white :/

**Comment by sonora** [comment]
Kust let me know when you want me to merge it, ok?

Yes, I share your view about recorded voices and do not put much effort in maintaining them other than making sure that updates to our config files do not break anything. Often, refinements we make for TTS can at the same time automatically improve the recorded voice, unless you need new vocabulary. In that case we simply need to live with that gap, some people provide a set of self-recordings, or we could use our script to produce ogg files from a license free synthesizer service. But I rarely do it for voices readily supported in TTS.

**Comment by akotynski** [comment]
Fixes can be merged. Let's skip `onto exit 6` improvement for now.

**Comment by sonora** [comment]
Ok, merged! I've also updated the recorded package, there also all prompts still produce sound -can't judge how good. 😉

**Comment by pm4rcin** [comment]
@akotynski I have tested today the version 4.3 with RHVoice and in my opinion https://github.com/osmandapp/OsmAnd-resources/blob/390fb0f8c68345d6166745a70ab4a34e05a9ffac/voice/pl/pl_tts.js#L115-L116 should be changed because not every TTS engine supports implicit declinations.

**Comment by pm4rcin** [comment]
@sonora also there's a problem that street names (and [destinations](https://wiki.openstreetmap.org/wiki/Key:destination)) doesn't have polish diactric marks like ę, ą, ś etc. **in OSMAnd** so announcing streets sounds strange with polish tts. Not sure if I should open separate issue?

**Comment by akotynski** [comment]
> @akotynski I have tested today the version 4.3 with RHVoice and in my opinion https://github.com/osmandapp/OsmAnd-resources/blob/390fb0f8c68345d6166745a70ab4a34e05a9ffac/voice/pl/pl_tts.js#L115-L116 should be changed because not every TTS engine supports implicit declinations.

Agree, the same logic as for kilometers can be applied.

> @sonora also there's a problem that street names (and [destinations](https://wiki.openstreetmap.org/wiki/Key:destination)) doesn't have polish diactric marks like ę, ą, ś etc. **in OSMAnd** so announcing streets sounds strange with polish tts. Not sure if I should open separate issue?

It probably needs more investigation, it looks like polish diactric marks are being displayed on the map correctly, so it's interesting where they get lost.

**Comment by sonora** [comment]
About the diacritic marks: Yes, please open a separate issue, it deserves a separate description and research.

Please note that the Development plugin's "Test voice prompt" feature has a setting where you can display (in writing on screen) all text passed to the TTS engine.

From what I have just tested, it does not look like any diacritics get lost at that stage. Chances are the engine converts all letters to canonical if e.g. your device or TTS engine is not set to the diacritic-specific locale. But not even that seems the case for me (tested with phone and device TTS set to American English, and pronouncing a Czech street name).
